src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameTests.cs

```
@@ -13,13 +13,14 @@ namespace System.Reflection.Metadata.Tests
     public class TypeNameTests
```

Is there test coverage for parsing a nested types with namespace e.g. `TypeName.Parse("a.b+c.d")`?

What is `TypeName.Parse("a.b+c.d")` going to return?

---

Updated to return `a` in your example.

---

> Passing System.Collections.Generic.Dictionary`2+MyRandomNamespace.Enumerator to Type.GetType returns the expected result

.NET Framework throws exception in this case. This looks like a bug in .NET Core - regression introduced by rewrite of the type name parser in C#. I do not think that the parser should be silently dropping any significant part of the input to the floor.

> Updated to return a in your example.

I think that the behavior for my example should be:
- Namespace returns "c", Name returns "d"
or
- Namespace returns empty string, Name returns "c.d"

---

I think the former option (Namespace returns `c`, Name does not change) makes more sense and will update the PR.

There's also the question of what `a.b+c.d+e` will return, and I think the answer should be `c` as well.

---

Done.

---

> a.b+c.d+e will return, and I think the answer should be c as well.

If it returns `c`, how does one distinguish between `a.b+c.d+e` and `a.b+c.d+c.e`?

I have been looking at handling of namespaces of nested types throughout the system. It seems that the system as a whole assumes that the nested types do not have namespaces. Would it best to avoid all handling of namespaces for nested types and treat the nested type name as a name without namespace even when it has a dot in it?

`a.b+c.d` -> Name: "c.d", Namespace: throws InvalidOperationException
`a.b+c.d+e` -> Name: "e", Namespace: throw InvalidOperationException

Alternatively, type.Namespace can return the outer most namespace instead of throwing to match reflection. However, it seems to be a pit of failure. Namespaces for nested types do not make sense. We have opted to throw InvalidOperationException in situations like this in the Typename APIs. For example, Typename.GetElementType throws InvalidOperationException, but the matching reflection API returns null.)

---

Updated `GetNamespace` to throw.

---

FWIW I am supportive of the "throw" decision for nested types. And I wish things like `a.b+c.d` were illegal ;)